### PR TITLE
Account details and snaps list

### PR DESCRIPTION
**Pull request by Lukewh**
Implements https://github.com/canonicalltd/snapcraft-design/issues/302
and https://github.com/canonicalltd/snapcraft-design/issues/292

# QA
- Pull the branch
- `./run`
- Visit http://0.0.0.0:8004/account it should redirect to /account/snaps
- Visit http://0.0.0.0:8004/account/snaps you should see your list of snaps
- Visit http://0.0.0.0:8004/account/details you should see your account information

To QA empty snaps page:
- edit `modules/publisher/views.py:101` and add:
```
snaps_list = {}
registered_snaps = {}
```

To QA a single snap:
- edit `snaps_list = {}` added above ^
```
snaps_list = {'name-of-snap': snaps_list['name-of-snap']}
```

Demo links:
http://snapcraft.io-pr-421.run.demo.haus/account
http://snapcraft.io-pr-421.run.demo.haus/account/details
http://snapcraft.io-pr-421.run.demo.haus/account/snaps

### /account/snaps No snaps

- [x] Snapcraft bird (though I imagine it will be the same outcome as with the 404 page - the bird shouldn't be used like that) @lyubomir-popov please check @Greggless wireframes: https://github.com/canonicalltd/snapcraft-design/issues/302

### /account/snaps First snap

- [ ] Data for 'You've just released your first snap to the testing "X" channel.' where X is the channel.
- [ ] Ability to toggle privacy of snap (conversation with @cprov needed/ started)
- [x] Up to date image of 'Market' page as will be on launch

### /account/snaps n snaps

- [x] Owner field - should be live in the API by 22-03-2018 (tomorrow at the time of writing)

### /account/details
screenshot-2018-4-19 account details linux software in the snap store

### /account/snaps No snaps
screenshot-2018-4-19 my snaps linux software in the snap store

### /account/snaps First snap
screenshot-2018-4-19 my snaps linux software in the snap store 4

### /account/snaps n snaps
screenshot-2018-4-19 my snaps linux software in the snap store 2

### /account/snaps Registered names
screenshot-2018-4-19 my snaps linux software in the snap store 3

### CODE REVIEW

**Comment by Greggless** [comment]
## Bug
- [ ] I'm getting the 'no snaps' view for /account/snaps, but I have indeed, got snaps
- [ ] dashboard API is not pulling through Ubuntu one username (are we waiting on store implementation here?)
image

## Errors

- [ ] URL should be: `/account/~~details~~` (we'll eventually have an `/account/settings` page which will be a child of `/account`)
- [ ] remove text from `/account/details` ' ...  For more information read the account documentation.' because there is no documentation for it...yet
- [ ] `/account/snaps` the text link 'publishing your first snap' should lead to: https://docs.snapcraft.io/build-snaps/your-first-snap
- [ ] No snapcraft logo on the 'no snaps' page, @lyubomir-popov can advise to it's usage (check [this issue](https://github.com/CanonicalLtd/snapcraft-design/issues/302#issuecomment-374181006) for reference)

## Changes based on latest API availability
### MVP table:

| Name | Privacy | Owner | 
| :- | - | - | 
| `Snap name` | `public` | `Display name` |

### The future table 👾 

| Name | Privacy | Owner | Latest revision | Version | Status | 
| :- | - | - | - | - | - | 
| `Snap name` | `Public` | `Display name` | `#N` & `channel` | `version_string` | `status` + `human friendly date & time` |

**Comment by Lukewh** [comment]
@Greggless for the MVP table 'Owner' column - should we say 'You' if the usernames match, or repeat the name?

**Comment by Lukewh** [comment]
I also feel *very* strongly that `/account` is the wrong URL for a cul-de-sac page. Account works for the current implementation of account details, followed by snap list as there's an obvious path to progress. People can and will easily guess `/account` and arrive at a dead end. 
`/account` should be redirected to `/account/snaps` or something more useful.
`account` is the box everything to do with your account goes in - snaps, user details, releases, builds etc.

We should maybe get some more input on this though as I could be wrong.

**Comment by Greggless** [comment]
**Owner**
+1 for if`snap_owner`=`current user` then `owner_value`=`you`

**re. `/account/details`:**
If you feel strongly about it then i'm not wedded to it being `/account`. We are using the URL structure with account as top level and all snap related things under that so it makes more sense for it to be `/account/details` in the end

**Comment by Greggless** [comment]
Loving your work! ❤️ 

### `/account/snaps`
<hr>

image

<hr>

**My snaps**
- [ ] Remove hr from under heading
- [ ] Muted heading for table headers

**Registered names**
- [ ] Remove snapcraft placeholder icon
- [ ] Put registered snaps in a comma separated list until we have more data points available in the API (We need to consult Vanilla rules about what headings we can use (Say H5) @kwm14 what say ye)

**Comment by Lukewh** [comment]
@Greggless I don't think 'Muted heading for table headers' is a thing in Vanilla

**Comment by kwm14** [comment]
@Greggless we actually discussed having 'Muted headings' for tables based on a proposal from @lyubomir-popov which can be found [here](https://github.com/vanilla-framework/vanilla-framework/issues/1621).

Me and @spencerbygraves reviewed and gave it a +1. Lyubo will be generating a PR for MAAS which will then be pushed up to Vanilla in a few weeks or so.

**Comment by Lukewh** [comment]
@kwm14 @Greggless cool, once they've landed I'll implement them on this page - which sounds like it should be before Berlin, so wunderbar!

**Comment by lyubomir-popov** [comment]
agree that the bird shouldn’t be used as in the mockup. The text just beeds a bit of styling, I’ll experiment and update the issue

**Comment by Greggless** [comment]
@lyubomir-popov I'm working on a more helpful empty state page, so I wouldn't worrying about the styling too much

**Comment by lyubomir-popov** [comment]
I agree there should be no snap logo on the 'no snaps' page.

**Comment by Lukewh** [comment]
@tbille you brute!

**Comment by Greggless** [comment]
# Enhancement for a user w/ no snaps
- If icons are too ambitious, we can ditch 'em
- Link to two different routes:
   - **1.** [Try the tutorial](https://tutorials.ubuntu.com/tutorial/create-your-first-snap#0)
   - **2.** [Choose a language guide](https://docs.snapcraft.io/build-snaps/languages)
- We can also link to the docs and forum 

image

File: templates/publisher/account-snaps.html [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+              <td>
+                {% if snaps[snap].publisher.username == current_user %}
+                  You
+                {% else %}
+                  {{ snaps[snap].publisher['display-name'] }}
+                {% endif %}
+              </td>
+              {% if snaps[snap].latest_revisions[0].status == 'Published' %}
+                <td>{{ snaps[snap].latest_revisions[0].channels[0] }}</td>
+              {% else %}}
```
Comment (bartaz) [inline templates/publisher/account-snaps.html]: Additional `}` leaking out of the table:

screen shot 2018-04-13 at 10 00 17

File: templates/publisher/account-snaps.html [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        </div>
+      {% endfor %}
+    {% endif %}
+  </div>
+{% endif %}
+{% if registered_snaps %}
+  <div class="p-strip is-shallow">
+    <div class="row">
+      <div class="col-12">
+        <h4 class="u-float--left">My registered snap names</h4>
```
Comment (bartaz) [inline templates/publisher/account-snaps.html]: Is `u-float--left` needed here? Is there anything else supposed to be here in this row for the float to be needed?

File: templates/publisher/account-snaps.html (hunk 1 above)
Comment (Lukewh) [reply templates/publisher/account-snaps.html]: I couldn't find that for the life of me, I thought @Greggless was lying! thanks.

File: templates/publisher/account-snaps.html (hunk 2 above)
Comment (Lukewh) [reply templates/publisher/account-snaps.html]: Good catch - in the original design there was a link to the right for registering a name, which has been removed. Fixed.

File: templates/publisher/account-snaps.html (hunk 1 above)
Comment (bartaz) [reply templates/publisher/account-snaps.html]: Rare occasion when `View source` is more useful then browser developer tools.

File: modules/publisher/views.py [hunk 3]
```diff
@@ -28,8 +28,52 @@ def refresh_redirect(path):
 
 
 def get_account():
```
[the remark this answers is not in the document]
Comment (Lukewh) [reply modules/publisher/views.py]: Done

File: templates/publisher/account-snaps.html [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+              {% endif %}
+              <td>{{ snaps[snap].latest_revisions[0].version }}</td>
+            </tr>
+          {% endfor %}
+        </tbody>
+      </table>
+    </div>
+    {% if snaps|length == 1 %}
+      {% for snap in snaps %}
+        <div class="p-strip--light is-shallow">
```
Comment (bartaz) [inline templates/publisher/account-snaps.html]: This is a p-strip inside another p-strip. It's a bit of a weird nesting and it causes that there is no spacing between the table and this strip.
screen shot 2018-04-13 at 11 31 55

Moving this strip one step higher in DOM tree fixes this issue:

screen shot 2018-04-13 at 11 33 02

File: templates/publisher/account-snaps.html [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+<div class="p-strip is-shallow u-no-padding--bottom">
+  <div class="row">
+    <div class="col-12">
+      <h1 class="p-heading--three u-float--left">My snaps</h1>
+      <a href="https://docs.snapcraft.io/build-snaps/" class="p-link--external u-float--right">How to create a snap</a>
+    </div>
+  </div>
+  <div class="row u-no-margin--top">
+    <div class="col-12">
+      <hr {% if not snaps %}class="u-no-margin--bottom" {% endif %}/>
```
Comment (bartaz) [inline templates/publisher/account-snaps.html]: Is this `hr` meant to be here at all when there is no snaps? It looks weird with light strip just below it:

screen shot 2018-04-13 at 11 26 01

**Comment by bartaz** [comment]
The content hierarchy for when there is only one snap seems a bit weird:

screen shot 2018-04-13 at 11 36 09

`Congratulations!` is H3 (but styled as h5 to look smaller), but `Market` (which is possibly lower in hierarchy) uses H4 and appears bigger.

Also `Things to do next:` seems empty. Unless the 'Market' section below is meant to be a thing to do next, but because of horizontal line it looks disconnected.

File: templates/publisher/account-snaps.html [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    <div class="row p-divide">
+      <div class="col-6 p-divider__block">
+        <h3 class="p-heading--five">Snapcraft documentation</h3>
+        <p>All you need to know about snaps, snapcraft and everything in between.</p>
+        <p><a href="https://docs.snapcraft.io">Read the docs &rsaquo;</a></p>
+      </div>
+      <div class="col-6 p-divider__block">
+        <h3 class="p-heading--five">Join the community</h3>
+        <p>Request features, troubleshoot and generally find all the latest talk about snaps, snapd and snapcraft.</p>
+        <p><a href="https://forum.snapcraft.io">Go to the focum &rsaquo;</a></p>
```
Comment (bartaz) [inline templates/publisher/account-snaps.html]: Typo: `focum`

**Comment by matthewpaulthomas** [comment]
Most of this work looks excellent to me. I’d make a couple of small deletions:
- “~~Things to do next:~~” — it’s unnecessary, and it’s misaligned with the block that it’s introducing.
- “Names you own but have not pushed a snap to ~~will appear in this list~~:”

I’m not a fan of either of the “no snaps” designs: they look _so_ different, from how people would expect a populated state to look, that people might easily think they’re in the wrong place. But that shouldn’t block this landing.

**Comment by lyubomir-popov** [comment]
@bartaz A few suggestions, please ignore if it's too late in the process to change.
Please wrap <b> in a block level element like a p, as b span etc are display:inline, and with the new vanilla, spacing is attached to elements, not * + *.
Not sure if a design was provided for this, but I would do something like this: 

image

1 deep strip, rows inside, helper text under value, not key. No hr below heading, just rely on the rows and * + * for spacing.

File: templates/publisher/account-snaps.html (hunk 6 above)
Comment (lyubomir-popov) [reply templates/publisher/account-snaps.html]: Is this going to have just one entry? If so, This layout is not appropriate. instead, Things to do next should be rephrased to: 
Next, would you like to improve listings in sotre? [   Go to Market   ]

**Comment by Greggless** [comment]
### Review
/account/detials - email address:
> Your email address will not be shared publically

**Comment by Greggless** [comment]
### /account/snaps

**Congratulations!** You've just released a snap! 🚀 
- Currently your snap is public `make in private now`
- Want to improve the listing in stores? You can add an icon, screenshots, and sections.
`[Go to market listing]`

**Comment by Lukewh** [comment]
It would be good to get this landed before we implement some of the fixes we need to do:

QA needed @lyubomir-popov @Greggless @bartaz

**Comment by Greggless** [comment]
Got any screenshots, babe?

**Comment by Greggless** [comment]
Will :+1:

**Comment by Lukewh** [comment]
@bartaz yeah, I want this landed so I can work against master to fix them :dancer:

**Comment by lyubomir-popov** [comment]
@Lukewh About this screen - 
screen shot 2018-04-19 at 10 53 46

Congratulations should be styled asp-heading--four, not --five.
Icon is on wrong background, transparent one here:
https://github.com/CanonicalLtd/snapcraft-design/pull/419

Can the hr be in the same col-12 ot reduce spacing a bit?

screen shot 2018-04-19 at 10 59 46

**Comment by Lukewh** [comment]
I've addressed the changes mentioned by @lyubomir-popov so going to merge

**Pull request merged**